pytest_girder/pytest_girder/utils.py

```diff
@@ -250,3 +312,107 @@ def buildHeaders(headers, cookie, user, token, basicAuth, authHeader):
+    used to send files with the request.
+
+    :param fields: List of (name, value) tuples.
+    :param files: List of (name, filename, content) tuples.
+    :param path: The path part of the URI.
+    :type path: str
+    :param method: The HTTP method.
+    :type method: str
+    :param prefix: The prefix to use before the path.
+    :param isJson: Whether the response is a JSON object.
+    :param token: Auth token to use.
+    :type token: str
+    :returns: The cherrypy response object from the request.
+    """
+    contentType, body, size = MultipartFormdataEncoder().encode(fields, files)
+
+    headers = [('Host', '127.0.0.1'),
+               ('Accept', 'application/json'),
+               ('Content-Type', contentType),
+               ('Content-Length', str(size))]
+
+    app = cherrypy.tree.apps['']
+    request, response = app.get_serving(local, remote, 'http', 'HTTP/1.1')
+    request.show_tracebacks = True
+
+    if token is not None:
+        headers.append(('Girder-Token', token))
+    elif user is not None:
+        headers.append(('Girder-Token', _genToken(user)))
+
+    fd = io.BytesIO(body)
+    # Python2 will not match Unicode URLs
+    url = str(prefix + path)
+    try:
+        response = request.run(method, url, None, 'HTTP/1.1', headers, fd)
+    finally:
+        fd.close()
+
+    if isJson:
+        body = getResponseBody(response)
+        try:
+            response.json = json.loads(body)
+        except ValueError:
+            raise AssertionError('Did not receive JSON response')
+
+    if response.output_status.startswith(b'500'):
+        raise AssertionError("Internal server error: %s" %
+                             getResponseBody(response))
+
+    return response
+
+
+def uploadFile(name, contents, user, parent, parentType='folder',
+               mimeType=None):
+    """
+    Upload a file. This is meant for small testing files, not very large
+    files that should be sent in multiple chunks.
+
+    :param name: The name of the file.
+    :type name: str
+    :param contents: The file contents
+    :type contents: str
+    :param user: The user performing the upload.
+    :type user: dict
+    :param parent: The parent document.
+    :type parent: dict
+    :param parentType: The type of the parent ("folder" or "item")
+    :type parentType: str
+    :param mimeType: Explicit MIME type to set on the file.
+    :type mimeType: str
+    :returns: The file that was created.
+    :rtype: dict
+    """
+    mimeType = mimeType or 'application/octet-stream'
+    resp = request(
+        path='/file', method='POST', user=user, params={
+            'parentType': parentType,
+            'parentId': str(parent['_id']),
+            'name': name,
+            'size': len(contents),
+            'mimeType': mimeType
+        })
+
+    fields = [('offset', 0), ('uploadId', resp.json['_id'])]
+    files = [('chunk', name, contents)]
+    resp = multipartRequest(
```

Comment:
It would be simpler to pass the data in the request body instead of using `multipart/form-data` encoding, especially considering this is intended only for small files. See: https://github.com/girder/girder/blob/6d9cc4d3cdbc13843f2eb74cca6aa08dd7649672/girder/api/v1/file.py#L235-L237

The upload could even be done in a single request by including the data in the `POST /file` request:
https://github.com/girder/girder/blob/6d9cc4d3cdbc13843f2eb74cca6aa08dd7649672/girder/api/v1/file.py#L77-L80

Reply:
@msmolens this utility is just for testing legacy multipart mode, which is deprecated. You're right that it shouldn't be used in production, but we do want to test it to make sure it works for legacy clients.

Reply:
If it is indeed for testing the legacy functionality, I think that should be noted on the fixture.

Reply:
Sorry it took me a long time to respond. I only copied and pasted code from the old test suite to the pytest_girder package. I am working on a plugin and I am using pytest_girder package. As part of my testing I wanted to test file uploads and I just copy-pasted functionality from the legacy test suite as I mentioned. Do you guys think it is better to solve issues you mentioned in a separate PR in the future?

Reply:
That makes sense. I think it would be useful to note in the commit message or PR description that the code is copied from the legacy test suite.

Reply:
@msmolens I added a note to the commit message. Let me know if that is descriptive enough.

Reply:
LGTM, and I'll defer to others for PR approval

Reply:
I think if we're going to implement only one mode in this helper, it should be the sanctioned mode (body + query params). I fear that otherwise, this code will not be touched for a long time and we'll forget that it's even using multipart internally. Using the normal upload mode also has the benefit of not duplicating arcane, ugly code, i.e. this PR will be much shorter and more palatable. Here's an example of uploading a chunk the modern way, taken from the old testing framework:

```
resp = self.request(
            path='/file/chunk', method='POST', user=self.user, body=chunk2, params={
                'offset': resp.json['offset'],
                'uploadId': uploadId
            }, type='text/plain')
```

The `body` param can be passed as an encoded string or a text type; if it's passed as a text type we will send it as UTF-8.

This change should be fairly straightforward. Let me know if you run into problems with it.